### PR TITLE
Dma interrupt

### PR DESCRIPTION
**Pull request by AndronikosKostas**

### CODE REVIEW

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 1]
```diff
@@ -1,34 +1,72 @@
...
+This callback will be executed when any of following events occurs :
+- HT (Half Transfer) : Half of Rx buffer is filled)
+- TC (Transfer Complete) : Rx buffer is full.
+(In case of Circular DMA, reception could go on, and next reception data will be stored
+    in index 0 of reception buffer by DMA).
+- Idle Event on Rx line : Triggered when RX line has been in idle state (normally high state)
+for 1 frame time, after last received byte.
+*/
+
+#define queue_length 10
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Move to `COMMS_Definitions`

Comment (AndronikosKostas) [reply Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Έκανα push again και έσβησα οτι είχε να κάνει με queues γιατί τώρα πια δε χρειάζονται

**Review by giatats — APPROVED** [review]

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 2]
```diff
@@ -1,34 +1,57 @@
...
-    uint8_t RxDmaBuffer[TcCommandSize];
-    etl::vector<uint8_t, TcCommandSize> TcCommand;
+    StackType_t taskStack[TaskStackDepth];
 
 public:
-
-    void execute();
-
+    // buffer that holds the data of the DMA //
+    // needs to be public in order the callback to have access to that //
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Do your comments like 

```
/**
 * Docs that say something
*/
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 3]
```diff
@@ -1,34 +1,57 @@
...
+    etl::vector<uint8_t, TcCommandSize> RxDmaBuffer;
+    // size of the incoming bytes from the UART //
+    uint16_t Size ;
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
    uint16_t incomingMessageSize ;
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 4]
```diff
@@ -1,34 +1,57 @@
...
+    StackType_t taskStack[TaskStackDepth];
 
 public:
-
-    void execute();
-
+    // buffer that holds the data of the DMA //
+    // needs to be public in order the callback to have access to that //
+    etl::vector<uint8_t, TcCommandSize> RxDmaBuffer;
+    // size of the incoming bytes from the UART //
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: fix docs like my comment above

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 5]
```diff
@@ -1,34 +1,57 @@
...
+    uint16_t Size ;
+    // constructor //
     TCHandlingTask() : Task("TCHandlingTask") {
-        HAL_UART_Receive_DMA(&huart3, RxDmaBuffer, DmaBufferSize);
+
     }
 
+    void execute();
+    // task creation //
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 6]
```diff
@@ -1,34 +1,57 @@
...
     void createTask() {
         xTaskCreateStatic(vClassTask < TCHandlingTask > , this->TaskName,
                           TCHandlingTask::TaskStackDepth, this, tskIDLE_PRIORITY + 1,
                           this->taskStack, &(this->taskBuffer));
-    }
 
+        HAL_UARTEx_ReceiveToIdle_DMA(&huart3, RxDmaBuffer.data(), TcCommandSize);
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Move this to class constructor

File: Core/Inc/Platform/COMMS_Definitions.hpp [hunk 7]
```diff
@@ -38,6 +38,7 @@ inline const uint8_t TcCommandSize = 64;
 inline const uint16_t DmaBufferSize = TcCommandSize;
 
 
+
```
Comment (Sourland) [inline Core/Inc/Platform/COMMS_Definitions.hpp]:
```suggestion
```

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 8]
```diff
@@ -1,42 +1,38 @@
...
-
-            uint8_t *ecssTCBytes = reinterpret_cast<uint8_t *>(cobsDecodedMsg.data());
-            auto ecssTC = MessageParser::parse(ecssTCBytes, messageLength);
-
-            LOG_DEBUG << "Received new TC[" << ecssTC.serviceType << "," << ecssTC.messageType << "]";
-            MessageParser::execute(ecssTC);
-
-        }
+    // TcCommand : the vector win which we copy the data from the RxDmaBuffer //
+    etl::vector<uint8_t, TcCommandSize> TcCommand;
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]: Add this as class property

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp:1]: Make sure no comments are present in the .cpp. The docs should be on .hpp files

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 9]
```diff
@@ -1,42 +1,38 @@
...
+    // Handle of the calling task //
+    taskHandle = xTaskGetCurrentTaskHandle();
+    // disabling the half buffer interrupt //
+    __HAL_DMA_DISABLE_IT(&hdma_usart3_rx, DMA_IT_HT);
+    // disabling the full buffer interrupt //
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]: add these after you initialize the uart in the constructor

File: Core/Src/main.cpp [hunk 10]
```diff
@@ -14,72 +14,63 @@
...
-
-void blinkyTask2(void * pvParameters){
-    for(;;){
-        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_0);
-        HAL_Delay(300);
-    }
-}
+// extern UART_HandleTypeDef huart3;
+// extern I2C_HandleTypeDef hi2c2;
+// extern RTC_HandleTypeDef hrtc;
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```

File: Core/Src/main.cpp [hunk 11]
```diff
@@ -14,72 +14,63 @@
...
-    canTestTask.emplace();
-    canGatekeeperTask.emplace();
-
+    // mcuTemperatureTask.emplace();
+    // temperatureSensorsTask.emplace();
+    // timeKeepingTask.emplace();
+    tcHandlingTask.emplace();
+    watchdogTask.emplace();
+    // canTestTask.emplace();
+    // canGatekeeperTask.emplace();
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
    mcuTemperatureTask.emplace();
    temperatureSensorsTask.emplace();
    timeKeepingTask.emplace();
    tcHandlingTask.emplace();
    watchdogTask.emplace();
    canTestTask.emplace();
    canGatekeeperTask.emplace();
```

File: Core/Src/main.cpp [hunk 12]
```diff
@@ -14,72 +14,63 @@
...
-//    tcHandlingTask->createTask();
-    canTestTask->createTask();
-    canGatekeeperTask->createTask();
+    // temperatureSensorsTask->createTask();
+    // mcuTemperatureTask->createTask();
+    // timeKeepingTask->createTask();
+    tcHandlingTask->createTask();
+    watchdogTask->createTask();
+    // canTestTask->createTask();
+    // canGatekeeperTask->createTask();
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
    temperatureSensorsTask->createTask();
    mcuTemperatureTask->createTask();
    timeKeepingTask->createTask();
    tcHandlingTask->createTask();
    watchdogTask->createTask();
    canTestTask->createTask();
    canGatekeeperTask->createTask();
```

File: Core/Src/main.cpp [hunk 13]
```diff
@@ -14,72 +14,63 @@
...
-    /**
-     * Uncomment below and comment above for Led task visualization (for STM32H743)
-     */
-//    xTaskCreate(blinkyTask1, "blinkyTask 2", 1000, nullptr, tskIDLE_PRIORITY + 1, nullptr);
-//    xTaskCreate(blinkyTask2, "blinkyTask 2", 1000, nullptr, tskIDLE_PRIORITY + 1, nullptr);
     for(;;);
     return;
 }
 
+// when this is called the UART is on idle mode, so the rxDmaPointer have some data //
```
Comment (Sourland) [inline Core/Src/main.cpp]: Add docs like this:
```
    /**
     * This function makes coffee and drinks it.
     * @param type the type of coffee to make
     * @return the coffee
     */
```

**Comment by Sourland** [comment]
check your code to follow this:
https://github.com/PeakSat-AUTH/obc-software/wiki/SpaceDot-Coding-Guidelines#documentation-and-comments

**Comment by AndronikosKostas** [comment]
Pushed again !!!

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 14]
```diff
@@ -1,34 +1,65 @@
...
+
 extern DMA_HandleTypeDef hdma_usart3_rx;
 extern UART_HandleTypeDef huart3;
 
 class TCHandlingTask : public Task {
 private:
-    const uint16_t delayMs = 1000;
+
+    /**
+     * stack depth for the freeRTOS
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
     * Stack depth for the freeRTOS
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 15]
```diff
@@ -1,34 +1,65 @@
...
+     */
     const static inline uint16_t TaskStackDepth = 1000;
-    StackType_t taskStack[TaskStackDepth];
 
-    uint8_t RxDmaBuffer[TcCommandSize];
+    StackType_t taskStack[TaskStackDepth];
+    /**
+    * TcCommand : buffer in which we copy the data from the RxDmaBuffer
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
    * Buffer in which we copy the data from the RxDmaBuffer
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 16]
```diff
@@ -1,34 +1,65 @@
...
+    */
     etl::vector<uint8_t, TcCommandSize> TcCommand;
 
 public:
-
-    void execute();
+    /**
+     *  RxDmaBuffer : buffer that holds the data of the DMA,  needs to be public in order the callback to have access to that
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
     *  Buffer that holds the data of the DMA,  needs to be public in order the callback to have access to that
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 17]
```diff
@@ -1,34 +1,65 @@
...
+     */
+    etl::vector<uint8_t, TcCommandSize> RxDmaBuffer;
+    /**
+     * size of the incoming bytes from the UART
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
     * Size of the incoming bytes from the UART
```

**Comment by AndronikosKostas** [comment]
Uncomment the other tasks to run also

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp (hunk 17 above)
Comment (AndronikosKostas) [reply Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: done

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp (hunk 16 above)
Comment (AndronikosKostas) [reply Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Done

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp (hunk 15 above)
Comment (AndronikosKostas) [reply Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]: Done

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 18]
```diff
@@ -1,34 +1,65 @@
...
+for 1 frame time, after last received byte.
+*/
+
 extern DMA_HandleTypeDef hdma_usart3_rx;
 extern UART_HandleTypeDef huart3;
 
 class TCHandlingTask : public Task {
 private:
-    const uint16_t delayMs = 1000;
+
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp]:
```suggestion
```

File: Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp [hunk 19]
```diff
@@ -1,34 +1,65 @@
...
     etl::vector<uint8_t, TcCommandSize> TcCommand;
 
 public:
-
-    void execute();
+    /**
+     *  Buffer that holds the data of the DMA,  needs to be public in order the callback to have access to that
+     */
+    etl::vector<uint8_t, TcCommandSize> RxDmaBuffer;
+    /**
```
Comment (Sourland) [inline Core/Inc/FreeRTOSTasks/TCHandlingTask.hpp:43]:
```suggestion

    /**
```

File: Core/Inc/Platform/COMMS_Definitions.hpp [hunk 20]
```diff
@@ -38,6 +38,7 @@ inline const uint8_t TcCommandSize = 64;
...
 /**
```
Comment (Sourland) [inline Core/Inc/Platform/COMMS_Definitions.hpp]:
```suggestion
inline const uint16_t DmaBufferSize = TcCommandSize;

/**
```

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 21]
```diff
@@ -1,42 +1,27 @@
...
-
-            uint8_t *ecssTCBytes = reinterpret_cast<uint8_t *>(cobsDecodedMsg.data());
-            auto ecssTC = MessageParser::parse(ecssTCBytes, messageLength);
-
-            LOG_DEBUG << "Received new TC[" << ecssTC.serviceType << "," << ecssTC.messageType << "]";
-            MessageParser::execute(ecssTC);
-
-        }
+    for(;;){
+        // Awaiting a notification indicating the arrival of data from the UART //
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]:
```suggestion
```

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 22]
```diff
@@ -1,42 +1,27 @@
...
+        xTaskNotifyWait(0, 0, nullptr, portMAX_DELAY);
+
+        for(uint8_t i = 0 ; i < tcHandlingTask->incomingMessageSize; i++)
+            TcCommand.push_back(tcHandlingTask->RxDmaBuffer[i]);
+        // Clear the RxDmaBuffer to receive the new data //
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]:
```suggestion
```

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 23]
```diff
@@ -1,42 +1,27 @@
...
+        tcHandlingTask->RxDmaBuffer.clear();
+
+        // Tc Parsing and Execution //
+        etl::string<TcCommandSize> cobsDecodedMsg = COBSdecode<TcCommandSize>(TcCommand.data(), tcHandlingTask->incomingMessageSize);
+        // decoded_size = incomingMessageSize - 1 //
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]:
```suggestion
```

File: Core/Src/FreeRTOSTasks/TCHandlingTask.cpp [hunk 24]
```diff
@@ -1,42 +1,27 @@
...
+    for(;;){
+        // Awaiting a notification indicating the arrival of data from the UART //
+        xTaskNotifyWait(0, 0, nullptr, portMAX_DELAY);
+
+        for(uint8_t i = 0 ; i < tcHandlingTask->incomingMessageSize; i++)
+            TcCommand.push_back(tcHandlingTask->RxDmaBuffer[i]);
+        // Clear the RxDmaBuffer to receive the new data //
+        tcHandlingTask->RxDmaBuffer.clear();
+
+        // Tc Parsing and Execution //
```
Comment (Sourland) [inline Core/Src/FreeRTOSTasks/TCHandlingTask.cpp]:
```suggestion
        // TC Parsing and Execution
```

File: Core/Src/main.cpp [hunk 25]
```diff
@@ -14,32 +14,14 @@
 #include "CAN.hpp"
 #include "stm32h7xx_hal_fdcan.h"
 #include "CANGatekeeperTask.hpp"
+#include "WatchdogTask.hpp"
 
 extern SPI_HandleTypeDef hspi1;
-extern UART_HandleTypeDef huart3;
-extern I2C_HandleTypeDef hi2c2;
-extern RTC_HandleTypeDef hrtc;
+// extern UART_HandleTypeDef huart3;
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```

File: Core/Src/main.cpp [hunk 26]
```diff
@@ -14,32 +14,14 @@
...
-
-void blinkyTask1(void * pvParameters){
-    for(;;){
-        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_0);
-        HAL_Delay(50);
-        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_14);
-        HAL_Delay(50);
-    }
-}
+// extern I2C_HandleTypeDef hi2c2;
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```

File: Core/Src/main.cpp [hunk 27]
```diff
@@ -14,32 +14,14 @@
...
 
-void blinkyTask2(void * pvParameters){
-    for(;;){
-        HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_0);
-        HAL_Delay(300);
-    }
-}
+// extern RTC_HandleTypeDef hrtc;
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```

File: Core/Src/main.cpp [hunk 28]
```diff
@@ -50,36 +32,56 @@ extern "C" void main_cpp(){
...
+/**
+ * HAL_UARTEx_RxEventCallback is a callback function for UART receiving which is enable by the
+ * the call of the function HAL_UARTEx_ReceiveToIdle_DMA().
+ * It is triggered whenever the UART line is IDLE.
+ * @param huart : pointer of type UART_HandleTypeDef
+ * @param Size : incoming size of data in bytes
+ */
+extern "C" void HAL_UARTEx_RxEventCallback(UART_HandleTypeDef *huart, uint16_t Size){
+    // Size is used for copying the right size of data to the TcCommand buffer,
+    // which belongs to the execute() of the tcHandlingTask
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```

Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
    // Size is used for copying the correct size of data to the TcCommand buffer,
    // of the TC Handling Task
```

File: Core/Src/main.cpp [hunk 29]
```diff
@@ -50,36 +32,56 @@ extern "C" void main_cpp(){
...
+    tcHandlingTask -> incomingMessageSize = Size;
+    BaseType_t xHigherPriorityTaskWoken;
+
+    // As always, xHigherPriorityTaskWoken is initialized to pdFALSE to be able to
+    // detect it getting set to pdTRUE inside an interrupt function
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
```
This is standard FreeRTOS behavior so no further explaination is required from our side.

File: Core/Src/main.cpp [hunk 30]
```diff
@@ -50,36 +32,56 @@ extern "C" void main_cpp(){
...
+    xHigherPriorityTaskWoken = pdFALSE;
+    xTaskNotifyFromISR(tcHandlingTask->taskHandle, 0, eNoAction,  &xHigherPriorityTaskWoken);
+    portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+
+    // start the DMA again to receive the next packet of data //
```
Comment (Sourland) [inline Core/Src/main.cpp]:
```suggestion
    // Reset the DMA to receive the next chunk of data
```

**Pull request merged**